### PR TITLE
feat: added support to install manually install plugins

### PR DESCRIPTION
**Pull request by philprime**
This adds the option to manually install Pulumi plugins.

We are currently using [this Sentry plugin](https://github.com/pulumiverse/pulumi-sentry) and repeatedly noticed the following issue:

```
Logging into the Pulumi Service backend.
pulumi preview on shared
  Error: code: -2
   stdout: 
   stderr: Command failed with exit code 255: pulumi preview --parallel 2147483647 --exec-agent pulumi/actions@v3 --color auto --exec-kind auto.local --event-log /tmp/automation-logs-preview-7OA2LT/eventlog.txt --stack shared --non-interactive
  error: could not load plugin for sentry provider 'urn:pulumi:shared::***-cloud::pulumi:providers:sentry::default_0_0_8_github_/api.github.com/pulumiverse': Could not automatically download and install resource plugin 'pulumi-resource-sentry' at version v0.0.8, install the plugin using `pulumi plugin install resource sentry v0.0.8`.
  Underlying error: error downloading plugin sentry to file: failed to download plugin: sentry-0.0.8: 403 HTTP error fetching plugin from https://get.pulumi.com/releases/plugins/pulumi-resource-sentry-v0.0.8-linux-amd64.tar.gz
  Previewing update (shared)
  
  View Live: https://app.pulumi.com/***/***-cloud/shared/previews/1e8ae06b-b2c3-49f0-b91f-db8ab01[35](https://github.com/kula-app/kula-mono-repo/actions/runs/3764563516/jobs/6399153978#step:11:36)54b

  @ Previewing update..........
```

As the log also states, as automatically installing the plugin failed, we are changing our workflow to manually install this specific plugin. Due to action re-installing the pulumi CLI in our Docker-based GitHub Actions, we want to install the plugins right before running other commands. *EDIT: after looking further into the code, I realized only `~/.pulumi/bin` is deleted by this step, and not the full pulumi installation folder*

Therefore this adds the option to declare a list of plugins to manually install.
As nested array objects are not supported by GitHub Actions, we decided to use a YAML-string and parse it at runtime instead.

### CODE REVIEW

**Comment by philprime** [comment]
Changing this to WIP as we have noticed issues with installing the plugins

**Comment by philprime** [comment]
After tinkering with the issue more, I realized that the main issue wasn't the installation from other sources, but the rate-limit of GitHub.
Therefore we fixed the issue on our end, by installing the plugin in another GitHub Action step:

```yaml
- # Re-install plugins
  name: Re-install Pulumi plugins
  working-directory: cloud/infra/plugins
  run: |
    pulumi plugin rm resource sentry --yes
    wget https://github.com/pulumiverse/pulumi-sentry/releases/download/v0.0.8/pulumi-resource-sentry-v0.0.8-linux-amd64.tar.gz
    pulumi plugin install resource \
      sentry v0.0.8 \
      --file pulumi-resource-sentry-v0.0.8-linux-amd64.tar.gz
```

Still if you want to add the option to install plugins from this step, I am happy to contribute

File: action.yml [hunk 1]
```diff
@@ -93,6 +93,17 @@ inputs:
     description: 'Colorize output. Choices are: always, never, raw, auto'
     required: false
     default: 'auto'
+  manual-plugins:
```
Comment (cobraz) [inline action.yml:96]: Maybe we should name this config key `plugins`?

I'm not sure we can call them _manual plugins_. AFAIK, They are plugins that don't have an install script, we need to manually install them.

File: package.json [hunk 2]
```diff
@@ -51,6 +52,7 @@
     "dedent": "^0.7.0",
     "envalid": "^7.3.1",
     "got": "^11.8.6",
+    "js-yaml": "^4.1.0",
```
Comment (cobraz) [inline package.json:55]: We already have `yaml` installed; would it make sense to use `yaml`?

If you feel `js-yaml` is an improvement, feel free to open a separate PR on that 👍

File: src/main.ts [hunk 3]
```diff
@@ -19,6 +19,7 @@ const main = async () => {
   core.debug('Configuration is loaded');
 
   await pulumiCli.downloadCli(config.options.pulumiVersion);
+  await pulumiCli.installManualPlugins(config.manualPlugins);
```
Comment (cobraz) [inline src/main.ts:22]:
```suggestion
	if (config.manualPlugins) {
  	await pulumiCli.installManualPlugins(config.manualPlugins);
  }
```

**Review by cobraz — COMMENTED** [review]
Thank you for this contribution! 🙏

I have left some notes that I would love for you to look at. I think this change really make sense 👍

**Comment by Moon1706** [comment]
Hello folks!
I think we can use the [build-in functionality](https://github.com/pulumi/pulumi/blob/master/sdk/nodejs/automation/workspace.ts#L181-L197) of [@pulumi/pulumi/automation](https://www.pulumi.com/docs/reference/pkg/nodejs/pulumi/pulumi/x/automation/). This option allows us to make code clearer.

[My suggestion](https://github.com/pulumi/actions/pull/812).

**Comment by philprime** [comment]
The build-in functionality seems cleaner to me too.

**Comment by philprime** [comment]
We are not using this functionality internally anymore, therefore I will close this PR now.
If you still want this implemented, let me know and I will apply the requested changes.

**Pull request closed without merging**